### PR TITLE
MAINT: fix numerical precision in nct.stats

### LINKED ISSUE #12919: BUG: numerical precision, use gammaln in nct.mean
possible precision improvement, no bug yet

I was looking at the mean for nct.
I didn't find any problems in my examples with nct mean, but the gamma values were very large 1e60.

using gammaln should improve precision
`np.exp(special.gammaln(df/2.-0.5) - special.gammaln(df/2.))`

```
df = 100
special.gamma(df/2.-0.5), special.gamma(df/2.)
(8.667601843135274e+61, 6.082818640342675e+62)
```

### PR DESCRIPTION
**Pull request by rlucas7**
#### Reference issue
Closes gh-12919

#### What does this implement/fix?
For values of df > 344 the stats method
    of nct would raise runtime warnings and
    become nan unnecessarily.

#### Additional information
There are other known numerical issues with `nct` that ***are not*** fixed by this PR. For example the `cdf` and `pdf` have numerical issues: 
cdf:
https://github.com/scipy/scipy/issues/8344
pdf:
https://github.com/scipy/scipy/issues/7104

It looks like both the `cdf` and `pdf` are (eventually) in fortran so I can't fix. 
Better to get this in and fix a known issue though so opening PR.

### CODE REVIEW

File: scipy/stats/_continuous_distns.py [hunk 1]
```diff
@@ -5872,7 +5872,7 @@ def _stats(self, df, nc, moments='mv'):
         #
         mu, mu2, g1, g2 = None, None, None, None
 
-        gfac = sc.gamma(df/2.-0.5) / sc.gamma(df/2.)
+        gfac = np.exp(sc.gammaln(df/2.-0.5) - sc.gammaln(df/2.))
```
Comment (pvanmulbregt) [inline scipy/stats/_continuous_distns.py:5875]: Does this change the precision of the calculation for smaller values of `df`?
`df` is an integer and `gamma(df/2)`, `gamma(df/2-0.5)` can, in principle, be computed exactly, so the original ratio computation maintains much of the precision of numerator and denominator.  Even if `sc.gammaln` is quite accurate, taking the `exp` in the proposed computation will tend to reduce the precision.

Comment (rlucas7) [reply scipy/stats/_continuous_distns.py:5875]: @pvanmulbregt this is a fair point to call out and in this case the actual difference is small, consider: 

```
>>> import scipy.special as sc 
>>> import numpy as np 
>>> df = np.arange(2, 344)
>>> new = np.exp(sc.gammaln(df/2.-0.5) - sc.gammaln(df/2.))
>>> old = sc.gamma(df/2.-0.5) / sc.gamma(df/2.)
>>> 
>>> import matplotlib.pyplot as plt
>>> plt.style.use('seaborn-whitegrid')
>>> diff = old - new
>>> rel_diff = (old - new) / old
>>> 
>>> max(np.abs(diff))
1.787459069646502e-14
>>> max(np.abs(rel_diff))
2.2741286535617743e-13
>>> 
```

and most of the differences are smaller than this at the lower `df` values. 

It doesn't look like github gives the option to include a figure but if you run the lines in listing above and then run

```
plt.plot(df, diff, '-')
plt.show()
```

and similar for `rel_diff` you'll see that it's really not until 150 or so for `df` that we start to see larger differences (for either `diff` or `rel_diff`). 

These seem like small enough numerical differences (e-13) that the numerical precision is not much of a concern, do you agree?

**Comment by josef-pkt** [comment]
Over the years we switch most usages of factorial and gamma to gammaln in cases like this in the distributions.
I never heard of a deterioration in precision.

I cannot think of a usecase where would get close to the lower extreme case df=1.

```
df = np.array([1+1e-13, 1+1e-8, 1+1e-4, 1.5, 2])
np.exp(sc.gammaln(df/2.-0.5) - sc.gammaln(df/2.)) - sc.gamma(df/2.-0.5) / sc.gamma(df/2.)
array([ 1.56250000e-02,  2.98023224e-08,  1.81898940e-12, -8.88178420e-16,
        0.00000000e+00])
```

(I have no idea if there is a noncentral Cauchy corner case, but mean will not exist at the limit df -> 1)

**Comment by josef-pkt** [comment]
In some cases using beta or betaln was a bit better than gammaln, but I never looked closely

```
np.exp(sc.gammaln(df/2.-0.5) - sc.gammaln(df/2.))
array([1.77245385, 1.12837917, 0.88622693])

np.exp(sc.betaln(df/2-0.5, 0.5) - sc.gammaln(0.5))
array([1.77245385, 1.12837917, 0.88622693])

sc.beta(df/2-0.5, 0.5) / sc.gamma(0.5)
array([1.77245385, 1.12837917, 0.88622693])
```

File: scipy/stats/_continuous_distns.py (hunk 1 above)
Comment (rlucas7) [reply scipy/stats/_continuous_distns.py:5875]: Looking at this again in the context of the `nct` the `df>2` in the `_stats` method otherwise the variance does not exist. The calculations above are a bit out of context in that regard so better to disregard the `df=2` value and assume `df>2`. To answer the question though @pvanmulbregt they do not degrade the numerical accuracy for ***valid*** smaller values of `df`.

Comment (josef-pkt) [reply scipy/stats/_continuous_distns.py:5875]: mean exists for df > 1, so df=2 is valid for the mean

**Comment by rlucas7** [comment]
> In some cases using beta or betaln was a bit better than gammaln, but I never looked closely
> 
> ```
> np.exp(sc.gammaln(df/2.-0.5) - sc.gammaln(df/2.))
> array([1.77245385, 1.12837917, 0.88622693])
> 
> np.exp(sc.betaln(df/2-0.5, 0.5) - sc.gammaln(0.5))
> array([1.77245385, 1.12837917, 0.88622693])
> 
> sc.beta(df/2-0.5, 0.5) / sc.gamma(0.5)
> array([1.77245385, 1.12837917, 0.88622693])
> ```

Good call out I hadn't thought of that one. Indeed looks like that has abs max diff smaller than the `gammaln` by a couple factors (mind you the difference is e-13 vs e-17 for `betaln` here) that is better accuracy so I've swapped out `gammaln` and put in `betaln` and subtracted the additional `gammaln` from the log beta. 

example: 
```
 >>> import scipy.special as sc 
>>> import numpy as np 
>>> df = np.arange(3, 344)
>>> new = np.exp(sc.gammaln(df/2.-0.5) - sc.gammaln(df/2.))
>>> old = sc.gamma(df/2.-0.5) / sc.gamma(df/2.)
>>> beta = np.exp(sc.betaln(df/2-0.5, 0.5) - sc.gammaln(0.5))
>>> diff = old - new
>>> diff_b = old - beta
>>> 
>>> max(np.abs(diff))
1.787459069646502e-14
>>> max(np.abs(diff_b))
5.551115123125783e-17
```
I rebased off master to fix the linter messages that were fixed there. 
With the force push I've swapped to the new `betaln` approach.

File: scipy/stats/_continuous_distns.py (hunk 1 above)
Comment (rlucas7) [reply scipy/stats/_continuous_distns.py]: @josef-pkt yes but in `nct.stats` the 'mv' (mean ***and*** variance) are the defaults

Comment (josef-pkt) [reply scipy/stats/_continuous_distns.py]: But in my original application, I only needed the mean.

I'm currently working with a case with chi2/ncx2 with df=1.6, so maybe I will run into an nct with df=1.6 eventually. :)

But in any case, precision looks good to me and we get a much larger valid range for large df

**Comment by rlucas7** [comment]
The single test failure is unrelated, this is the pttrs linalg test that is failing: 

```
C:\hostedtoolcache\windows\Python\3.6.8\x86\lib\site-packages\scipy\linalg\tests\test_lapack.py:2191: in test_pttrf_pttrs
    b = A@x
E   RuntimeWarning: invalid value encountered in matmul
        A          = array([[4.3745403 +0.j        , 0.02058449-0.29122913j,
        0.        +0.j        , 0.        +0.j        ,
      ...       +0.j        , 0.        +0.j        ,
        0.43194503+0.5142344j , 4.7080727 +0.j        ]], dtype=complex64)
        D          = array([[4.3745403, 0.       , 0.       , 0.       , 0.       , 0.       ,
        0.       , 0.       , 0.       , 0. ...     , 0.       , 0.       , 0.       , 0.       ,
        0.       , 0.       , 0.       , 4.6085963]], dtype=float32)
        L          = array([[1.        +0.j        , 0.        +0.j        ,
        0.        +0.j        , 0.        +0.j        ,
      ...    ,
        0.        +0.j        , 0.        +0.j        ,
        0.09527042+0.11342029j, 1.        +0.j        ]])
        _d         = array([4.3745403, 4.931229 , 4.465308 , 4.439113 , 4.126635 , 4.1154575,
       3.999369 , 4.688882 , 4.533884 , 4.6085963], dtype=float32)
        _e         = array([0.00470552+0.06657365j, 0.19668724+0.12407716j,
       0.18642445+0.03123947j, 0.04783368+0.06581149j,
       0...79+0.11081878j,
       0.07607257+0.19632497j, 0.11191504+0.04258452j,
       0.09527042+0.11342029j], dtype=complex64)
        atol       = 1.1920928955078125e-05
        d          = array([4.3745403, 4.950714 , 4.7319937, 4.5986586, 4.1560187, 4.1559944,
       4.0580835, 4.866176 , 4.601115 , 4.7080727], dtype=float32)
        ddtype     = <class 'numpy.float32'>
        diag_cpy   = [array([4.3745403, 4.950714 , 4.7319937, 4.5986586, 4.1560187, 4.1559944,
       4.0580835, 4.866176 , 4.601115 , 4.70... +0.45606998j,
       0.30424225+0.785176j  , 0.52475643+0.19967379j,
       0.43194503+0.5142344j ], dtype=complex64)]
        dtype      = <class 'numpy.complex64'>
        e          = array([0.02058449+0.29122913j, 0.96990985+0.6118529j ,
       0.83244264+0.13949387j, 0.21233912+0.29214466j,
       0...5 +0.45606998j,
       0.30424225+0.785176j  , 0.52475643+0.19967379j,
       0.43194503+0.5142344j ], dtype=complex64)
        info       = 0
        n          = 10
        pttrf      = <fortran object>
        x          = array([0.59241456+0.684233j  , 0.04645041+0.4401525j ,
       0.60754484+0.12203824j, 0.17052412+0.4951769j ,
       0....965632  +0.25877997j, 0.80839735+0.66252226j,
       0.30461377+0.31171107j, 0.09767211+0.52006805j], dtype=complex64)
```

File: scipy/stats/_continuous_distns.py (hunk 1 above)
Comment (rlucas7) [reply scipy/stats/_continuous_distns.py]: > But in my original application, I only needed the mean.
> 
> I'm currently working with a case with chi2/ncx2 with df=1.6, so maybe I will run into an nct with df=1.6 eventually. :)

Yeah the only potential issue I see is:

https://github.com/scipy/scipy/blob/master/scipy/stats/_continuous_distns.py#L5877

which assumes you're `df>2`, for the `mean` alone (`_stats` is in the callpath of `mean`) this isn't an issue but if `df=2` you'll get a runtime error, even in the mean. This is why I'm  saying 'assumes `df>2`'. 
 
> But in any case, precision looks good to me and we get a much larger valid range for large df

Ok. I think if you've got statsmodels applications w/`df=1.6` then they could have `df=2` or something thereabouts, right? 

I'm thinking we can just wrap `c20` and `c22` calculations in an `if df > 2` block, seems like a simple enough fix to remove the potential runtime warning.

Comment (josef-pkt) [reply scipy/stats/_continuous_distns.py]: `if df > 2` doesn't work because it needs to be vectorized
similar to the where cases

**Comment by josef-pkt** [comment]
BTW: `mu = np.where(df > 1, nc*c11, np.inf)`
looks wrong to me, mu should be `nan` (undefined) instead of `inf`

I didn't check the reference in the code comment, Wikipedia says undefined

`t_gen` also has inf mean mu, which I think is also wrong
but t_gen has three cases for mu2 (variance) nan, inf and finite which looks correct to me.

**Comment by rlucas7** [comment]
Ah yes, vectorization.

We can put the c20 calculation into a np.where() with inf as value if df<2 that should work. 

I didn’t notice the inf part. I agree it is not inf, they have a small Wikipedia section explaining:

https://en.m.wikipedia.org/wiki/Cauchy_distribution#Mean

I usually think about it via contour integrals, different approach but you come to the same conclusion.

Note this is a backwards incompatible change. 

It is a defect so we should change it here, agreed? 

Sincerely, 

-Lucas Roberts

**Comment by rlucas7** [comment]
> BTW: `mu = np.where(df > 1, nc*c11, np.inf)`
> looks wrong to me, mu should be `nan` (undefined) instead of `inf`
> 
> I didn't check the reference in the code comment, Wikipedia says undefined
> 
> `t_gen` also has inf mean mu, which I think is also wrong
> but t_gen has three cases for mu2 (variance) nan, inf and finite which looks correct to me.

I added changes for the `mu` and `mu2` to be `nan` and updated a test to account for edge cases. 
The `df` argument to `_stats` still supports vectorization, I used the `np.where`

I did not change the `t_gen` that is a different distribution so I'd prefer to follow a separate PR to take a look at any issues with that distribution.

**Comment by rlucas7** [comment]
@josef-pkt @pvanmulbregt this PR is ready for review now, CI is green and fixes original issue as well as some newly identified ones.

File: scipy/stats/tests/test_distributions.py [hunk 2]
```diff
@@ -3373,14 +3373,32 @@ def test_variance_gh_issue_2401(self):
...
         assert_equal(k, np.nan)
 
+    def test_nct_stats_large_df_values(self):
+        # previously gamma function was used which lost precision at df=345
+        # cf. https://github.com/scipy/scipy/issues/12919 for details
+        nct_mean_df_1000 = stats.nct.mean(1000, 2)
+        nct_stats_df_1000 = stats.nct.stats(1000, 2)
+        expected_stats_df_1000 = [np.array(2.00150156), np.array(1.00401153)]
+        assert_allclose(nct_mean_df_1000, 2.0015015641427127)
+        assert_allclose(nct_stats_df_1000, expected_stats_df_1000)
```
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_distributions.py:3394]: More precise expected values:
```suggestion
        # These expected values were computed with mpmath. They were also
        # verified with the Wolfram Alpha expressions:
        #     Mean[NoncentralStudentTDistribution[1000, 2]]
        #     Var[NoncentralStudentTDistribution[1000, 2]]
        expected_stats_df_1000 = [2.0015015641422464, 1.0040115288163005]
        assert_allclose(nct_mean_df_1000, expected_stats_df_1000[0],
                        rtol=1e-13)
        assert_allclose(nct_stats_df_1000, expected_stats_df_1000,
                        rtol=1e-13)
```

File: scipy/stats/tests/test_distributions.py [hunk 3]
```diff
@@ -3373,14 +3373,32 @@ def test_variance_gh_issue_2401(self):
...
+        # and a bigger df value
+        nct_mean = stats.nct.mean(100000, 2)
+        nct_stats = stats.nct.stats(100000, 2)
+        expected_stats = [np.array(2.000015000062472), np.array(1.00004)]
+        assert_allclose(nct_mean, 2.000015000062472)
+        assert_allclose(nct_stats, expected_stats)
```
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_distributions.py:3400]: More precise expectd values:
```suggestion
        # These expected values were computed with mpmath.
        expected_stats = [2.0000150001562518, 1.0000400011500288]
        assert_allclose(nct_mean, expected_stats[0], rtol=1e-10)
        assert_allclose(nct_stats, expected_stats, rtol=1e-10)
```

**Comment by WarrenWeckesser** [comment]
@rlucas7, looks good.  I suggested a couple edits that provide more precise expected values for the unit tests.  We might as well use full precision expected values when we can.

File: scipy/stats/tests/test_distributions.py (hunk 2 above)
Comment (rlucas7) [reply scipy/stats/tests/test_distributions.py:3394]: thanks for the extra precision @WarrenWeckesser I always forget to `mpmath` things :)

**Comment by rlucas7** [comment]
> @rlucas7, looks good. I suggested a couple edits that provide more precise expected values for the unit tests. We might as well use full precision expected values when we can.

I committed the 2 suggestions @WarrenWeckesser thanks!

**Comment by WarrenWeckesser** [comment]
Apparently using `rtol=1e-13` in the unit test was too optimistic (but it did pass for me locally).  We can try bumping that up to 1e-12.

**Comment by rlucas7** [comment]
Seems like the 3.9 build on CI is failing b/c the suggested changes push it just past the `rtol` setting:

https://github.com/scipy/scipy/pull/12936/checks?check_run_id=1377694902#step:6:711

I also confirm the difference locally on 3.6 
```
>>> nct_mean_df_1000 = 2.0015015641427127
>>> expected = 2.0015015641422464
>>> nct_mean_df_1000 / expected 
1.000000000000233
>>> (nct_mean_df_1000 - expected) / expected 
2.3297192402764784e-13
>>> 
```
and the test itself passes when I rerun the test suite:
```
(scipy-dev) Lucass-MacBook:scipy rlucas$ python runtests.py -v -s stats 

... output elided ...
scipy/stats/tests/test_stats.py::TestMGCErrorWarnings::test_error_infty PASSED                                                                                                [ 99%]
scipy/stats/tests/test_tukeylambda_stats.py::test_tukeylambda_stats_known_exact PASSED                                                                                        [ 99%]
scipy/stats/tests/test_tukeylambda_stats.py::test_tukeylambda_stats_mpmath PASSED                                                                                             [ 99%]
scipy/stats/tests/test_tukeylambda_stats.py::test_tukeylambda_stats_invalid PASSED                                                                                            [100%]

====================================================== 3857 passed, 685 skipped, 264 deselected, 2 xfailed in 74.32s (0:01:14) ======================================================
(scipy-dev) Lucass-MacBook:scipy rlucas$ 
```
I can try to move `rtol` a bit to get this to pass, but not sure if the issue is that or something else.

**Comment by rlucas7** [comment]
> Apparently using `rtol=1e-13` in the unit test was too optimistic (but it did pass for me locally). We can try bumping that up to 1e-12.

Just seeing this now, yeah I see the same thing on my mac too, everything passes. The failures do seem to be consistent across python versions, platforms etc :) 

I'll change out the rtol to 1e-12 and see if that gets it green.

**Comment by rlucas7** [comment]
Even `rtol=1e-10` doesn't seem to fix it-not sure what the issue is but is too late tonight. 
I'll take a look again tomorrow.

File: scipy/stats/tests/test_distributions.py [hunk 4]
```diff
@@ -3429,14 +3429,39 @@ def test_variance_gh_issue_2401(self):
...
+                        rtol=1e-10)
+        assert_allclose(nct_stats_df_1000, expected_stats_df_1000,
+                        rtol=1e-10)
+        # and a bigger df value
+        nct_mean = stats.nct.mean(100000, 2)
+        nct_stats = stats.nct.stats(100000, 2)
+        # These expected values were computed with mpmath.
+        expected_stats = [2.0000150001562518, 1.0000400011500288]
+        assert_allclose(nct_mean, expected_stats[0], rtol=1e-10)
+        assert_allclose(nct_stats, expected_stats, rtol=1e-10)
```
Comment (WarrenWeckesser) [inline scipy/stats/tests/test_distributions.py:3463]: This is the test that is failing, which means it is the second component of `nct_stats` (i.e. the variance) that has the larger relative error.  The tests should pass if we bump up `rtol` to 1e-9.
```suggestion
        assert_allclose(nct_stats, expected_stats, rtol=1e-9)
```

Comment (rlucas7) [reply scipy/stats/tests/test_distributions.py]: > This is the test that is failing, which means it is the second component of `nct_stats` (i.e. the variance) that has the larger relative error. The tests should pass if we bump up `rtol` to 1e-9.

Ah, ok. This explains then, the previous assertion failure was the line just prior that tests on the mean. 
This means a bump to `rtol=1e-10` fixed that assertion. 

I committed the suggestion on the `rtol=1e-9` here for the variance, hopefully that is the last of the precision failing tests.

**Comment by rlucas7** [comment]
> Even `rtol=1e-10` doesn't seem to fix it-not sure what the issue is but is too late tonight.
> I'll take a look again tomorrow.
> 
> https://github.com/scipy/scipy/pull/12936/checks?check_run_id=1377985563

So to clarify my comment from last night, the `rtol=1e-10` makes the assertion: 

```
assert_allclose(nct_mean_df_1000, expected_stats_df_1000[0], rtol=1e-10)
```
pass. What Warren noted is that the variance in the second assertion was also failing:
```
assert_allclose(nct_stats_df_1000, expected_stats_df_1000, rtol=1e-10)
```
and it does look like moving `rtol=1e-9` in this second assertion should get the tests to green. 
This `rtol=1e-9` is the changes that are running now.

**Comment by rlucas7** [comment]
@WarrenWeckesser looks like the new `rtol` settings for mean and variance are green now

**Comment by WarrenWeckesser** [comment]
Thanks @rlucas7, merged.
